modopt/tests/test_signal.py

```diff
@@ -1,322 +1,233 @@
+    data2 = np.array(
+        [[0, 2.0, 2.0], [4.0, 5.0, 10], [11.0, 15.0, 18.0]],
+    )
+    data3 = np.array(
+        [
             [1.62434536, 0.38824359, 1.47182825],
             [1.92703138, 4.86540763, 2.6984613],
             [7.74481176, 6.2387931, 8.3190391],
-        ])
-        self.data4 = np.array([[0, 0, 0], [0, 0, 5.0], [6.0, 7.0, 8.0]])
-        self.data5 = np.array(
-            [[0, 0, 0], [0, 0, 0], [1.0, 2.0, 3.0]],
-        )
-
-    def tearDown(self):
-        """Unset test parameter values."""
-        self.data1 = None
-        self.data2 = None
-        self.data3 = None
-        self.data4 = None
-        self.data5 = None
-
-    def test_add_noise_poisson(self):
-        """Test add_noise with Poisson noise."""
-        np.random.seed(1)
-        npt.assert_array_equal(
-            noise.add_noise(self.data1, noise_type='poisson'),
-            self.data2,
-            err_msg='Incorrect noise: Poisson',
-        )
-
-        npt.assert_raises(
-            ValueError,
-            noise.add_noise,
-            self.data1,
-            noise_type='bla',
-        )
-
-        npt.assert_raises(ValueError, noise.add_noise, self.data1, (1, 1))
-
-    def test_add_noise_gaussian(self):
-        """Test add_noise with Gaussian noise."""
+        ]
+    )
+    data4 = np.array([[0, 0, 0], [0, 0, 5.0], [6.0, 7.0, 8.0]])
+    data5 = np.array(
+        [[0, 0, 0], [0, 0, 0], [1.0, 2.0, 3.0]],
+    )
+
+    @pytest.mark.parametrize(
+        ("data", "noise_type", "sigma", "data_noise"),
+        [
+            (data1, "poisson", 1, data2),
+            (data1, "gauss", 1, data3),
+            (data1, "gauss", (1, 1, 1), data3),
+            failparam(data1, "fail", 1, data1),
+        ],
+    )
+    def test_add_noise(self, data, noise_type, sigma, data_noise):
+        """Test add_noise."""
         np.random.seed(1)
         npt.assert_almost_equal(
-            noise.add_noise(self.data1),
-            self.data3,
-            err_msg='Incorrect noise: Gaussian',
+            noise.add_noise(data, sigma=sigma, noise_type=noise_type), data_noise
         )
 
-        np.random.seed(1)
-        npt.assert_almost_equal(
-            noise.add_noise(self.data1, sigma=(1, 1, 1)),
-            self.data3,
-            err_msg='Incorrect noise: Gaussian',
-        )
-
-    def test_thresh_hard(self):
-        """Test thresh with hard threshold."""
-        npt.assert_array_equal(
-            noise.thresh(self.data1, 5),
-            self.data4,
-            err_msg='Incorrect threshold: hard',
-        )
-
-        npt.assert_raises(
-            ValueError,
-            noise.thresh,
-            self.data1,
-            5,
-            threshold_type='bla',
-        )
-
-    def test_thresh_soft(self):
-        """Test thresh with soft threshold."""
+    @pytest.mark.parametrize(
+        ("threshold_type", "result"),
+        [("hard", data4), ("soft", data5), failparam("fail", None, raises=ValueError)],
```

Comment:
Isn't `ValueError` the default value for `raises` in `failparam`? I think we should just be consistent in either always explicitly declaring the error type or only when it is different from `ValueError`.

Reply:
I agree with that, lets make the specification of raise Type mandatory.